circuit-macros/src/circuit_type/proof_linking.rs

```diff
@@ -0,0 +1,163 @@
+
+    let mut fields_expr: Punctuated<FieldValue, Comma> = Punctuated::new();
+    for field in base_type.fields.iter().cloned() {
+        let ident = field.ident.clone().expect("only named fields are supported");
+        let field_type = field.ty.clone();
+
+        // This block expression represents the allocation of the current field into the
+        // constraint system
+        let group_expr = parse_linking_groups_for_field(&field);
+        let field_expr: Expr = parse_quote! {{
+            let groups = #group_expr;
+            let vars = self.#ident
+                .to_scalars()
+                .into_iter()
+                .map(|s| cs.create_variable_with_link_groups(s.inner(), &groups).unwrap())
+                .collect::<Vec<_>>();
+
+            <#field_type as #circuit_base_type>::#var_type::from_vars(&mut vars.into_iter(), cs)
+        }};
+
+        fields_expr.push(FieldValue {
+            attrs: Vec::new(),
+            member: Member::Named(ident),
+            colon_token: Some(Colon::default()),
+            expr: field_expr,
+        });
+    }
+
+    let method_name = new_ident(CREATE_WITNESS_METHOD_NAME);
+    let var_type = new_ident(VAR_TYPE_ASSOCIATED_NAME);
+
+    quote! {
+        fn #method_name(&self, cs: &mut PlonkCircuit) -> Self::#var_type {
+            Self::#var_type {
+                #fields_expr
+            }
+        }
+    }
+}
+
+/// Returns whether an attribute is a `linking_group_ids` attribute
+fn is_link_group_attr(attr: &Attribute) -> bool {
+    // Search the path segments
+    for seg in attr.path.segments.iter() {
+        if seg.ident == LINKING_GROUP_IDS_ATTR {
+            return true;
+        }
+    }
+
+    false
+}
+
+/// Parse a linking group expr of the form `["group1", "group2", ...]`
+///
+/// This allows us to quickly assign a field into groups
+fn parse_linking_groups_for_field(field: &Field) -> Expr {
+    for attr in field.attrs.iter() {
+        if is_link_group_attr(attr) {
+            return parse_linking_groups(attr);
+        }
+    }
+
+    // If no attribute was found, the field is a member of no link groups
+    parse_quote! { [] }
+}
+
+/// Parse a list of groups from a `link_groups` attribute
+fn parse_linking_groups(attr: &Attribute) -> Expr {
+    // Expected format is an "=" followed by a string literal that we can parse
+    // directly, i.e. `link_groups = "group1, group2, group3"`
+    let mut token_iter = attr.tokens.clone().into_iter();
+
+    // Validate that the next token is an "="
+    let next = token_iter.next().expect("expected '='");
+    match next {
+        TokenTree::Punct(punct) => {
+            assert_eq!(punct.as_char(), '=', "expected '='");
+        },
+        _ => panic!("expected '='"),
+    }
+
+    // Pull a string literal and parse groups from this directly
+    let groups = match token_iter.next().expect("expected string literal") {
+        TokenTree::Literal(lit) => parse_linking_groups_from_str(&lit),
```

Comment:
If we expect the attribute to contain a string literal, this won't exactly play well with using constantized link group names as in the last PR, will it?

Probably not worth the complexity to support constants, but just calling it out

Reply:
Correct, that's not the intent, you register a link group with the circuit and refer to it in the witness definition